File: src/version-changer/gradle-properties.ts
import * as fs from 'fs'
import {PropertiesFile} from '../files/properties'

import {ChangerDescriptor, VersionChanger} from '.'

export class GradleProperties implements VersionChanger {
  private readonly property: string
  private readonly path: string

  static createFromDesc({info: property, path}: ChangerDescriptor): GradleProperties {
    return new GradleProperties({
      property: property || 'version',
      path: path || 'gradle.properties',
    })
  }

  private constructor(arg: {property?: string; path?: string}) {
    this.property = arg.property ?? 'version'
    this.path = arg.path ?? 'gradle.properties'
  }

  async loadVersion(): Promise<string> {
    const source = await fs.promises.readFile(this.path, { encoding: 'utf-8' })
    const properties = PropertiesFile.parse(source)
    const version = properties.get(this.property)
    if (!version)
      throw new Error(`no such property: ${this.property}`)
    return version
  }

  async setVersion(version: string): Promise<void> {
    const source = await fs.promises.readFile(this.path, { encoding: 'utf-8' })
    const properties = PropertiesFile.parse(source)
    properties.set(this.property, version)
    await fs.promises.writeFile(this.path, properties.toSource(), {encoding: 'utf-8'})
  }

  toString(): string {
    return `gradle-properties(at ${this.path} prop ${this.property})`
  }
}

File: src/version_changer/gradle_properties.rs
use std::fmt::{Display, Formatter};
use std::path::PathBuf;
use serde::Deserialize;
use crate::utils::properties::PropertiesFile;
use crate::version_changer::VersionChanger;

#[derive(Debug, Deserialize)]
pub struct GradleProperties {
    #[serde(default = "path_default")]
    path: PathBuf,
    #[serde(alias="info", default = "default_property")]
    property: String,
}

fn path_default() -> PathBuf {
    PathBuf::from("gradle.properties")
}

fn default_property() -> String {
    "version".to_string()
}

impl Display for GradleProperties {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "gradle-properties(at {} prop {})", self.path.display(), self.property)
    }
}

impl VersionChanger for GradleProperties {
    fn parse(info: &str, path: &str) -> Self {
        Self {
            path: if path.is_empty() {
                path_default()
            } else {
                PathBuf::from(path)
            },
            property: if info.is_empty() {
                default_property()
            } else {
                info.to_string()
            },
        }
    }

    async fn load_version(&self) -> String {
        PropertiesFile::load_may_not_exist(&self.path)
            .await
            .expect("loading gradle.properties")
            .get(&self.property)
            .expect("getting version from gradle.properties")
    }

    async fn set_version(&self, version: &str) {
        let mut properties = PropertiesFile::load_may_not_exist(&self.path)
            .await
            .expect("loading gradle.properties");
        properties.set(&self.property, version.to_string());
        tokio::fs::write(&self.path, properties.to_string())
            .await
            .expect("writing gradle.properties");
    }
}
